common/mclock/simclock.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2016 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// package mclock is a wrapper for a monotonic clock source
+package mclock
+
+import (
+	"runtime"
+	"sync"
+	"time"
+)
+
+type event struct {
+	do func()
+	at AbsTime
+}
+
+// SimulatedClock implements a virtual Clock for reproducible time-sensitive tests.
+// It simulates a scheduler on a virtual timescale where actual processing takes zero time.
+//
+// Note: since there is no way in Go to know when all goroutines have reached a waiting
+// state (which should theoretically happen in each virtual moment), the algorithm runs
+// GoSched a fixed number of times after each step and limits time steps in order to
+// minimize precision loss (see maxStep and goSchedCount).
+type SimulatedClock struct {
+	now       AbsTime
+	scheduled []event
+	stop      bool
+	lock      sync.RWMutex
+}
+
+// NewSimulatedClock creates a new simulated clock
+func NewSimulatedClock(maxStep time.Duration, goSchedCount int) *SimulatedClock {
+	s := &SimulatedClock{scheduled: make([]event, 0, 100)}
+
+	go func() {
+		lastScheduled := 0
+		for {
+			for i := 0; i < goSchedCount; i++ {
+				runtime.Gosched()
```

Comment:
This loop is weird. You should never make assumptions about the scheduler, particularly when it comes to sleeping goroutines. You need very good justification to use `runtime.Gosched`.

An acceptable use of `Gosched` would be to when your goroutine is running a tight loop that cannot be preempted because it contains no function calls or channel synchronization. In such cases, the call to `Gosched` can ensure that other goroutines will run.

Reply:
I admit this is a hacky solution but it is only used for testing and it proved to be really useful. It is also used in https://github.com/ethereum/go-ethereum/pull/16490 which would be practically impossible to properly test without using a virtual timescale. This is the most reliable and least ugly solution I could come up with but I am open to suggestions.

Reply:
I understand that you need a virtual clock, my only issue is the call to `Gosched`. Does the test work without this call?  

Reply:
No. We need something to let goroutines run until they stop waiting on something again. A Sleep works too but it procuces less accurate and less reliable/reproducible results and also I'm not sure if it is less ugly.

Reply:
I've verified this now and the tests do pass without the `Gosched` call, they're just slower.